### PR TITLE
Using esbuild instead of rollup

### PR DESCRIPTION
**Pull request by WebReflection**
### Description

This PR removes *rollup* in favor of *esbuild*, ensuring `DEBUG` flag, when used, is available to the produced JS and the *umd* dance has been simplified, ditching the *amd* part from it as I think *amd* is kinda dead as module system.

### CODE REVIEW

**Comment by hoodmane** [comment]
Thanks for working on this @WebReflection!

>  * apparently esbuild deosn't support *umd* as output format ... it's currently just an *iife* ... would that work?

I am not sure. I think we don't really need `umd`: what `pyodide.js` does is it imports `loadPyodide` from `pyodide.mjs` and then sets `globalThis.loadPyodide = loadPyodide`. It is primarily intended for `<script src="pyodide.js">`, if you are using a bundler or loading code from an es6 module you are encouraged to you `pyodide.mjs`. So I think IIFE is fine. 
 
>  * apparently the *tty* module was unknown ... I've made it external but I am not sure that's the right thing to do

This is the correct thing to do.

>   * apparently we were converting async iterators to ES2017 as target so I had to raise the target to ES2018 ... is that OK?

I think this is fine. Our target is not very well considered. It was last modified by @jasongrout at Pycon IIRC...

**Comment by hoodmane** [comment]
It looks like the output has been name mangling which [we have not been doing](https://github.com/pyodide/pyodide/blob/main/src/js/rollup.config.mjs#L45) so to avoid changing things it would be good if we could turn that off.

**Comment by hoodmane** [comment]
Before we had:
```js
              global_defs: {
                DEBUG,
              },
```
https://github.com/pyodide/pyodide/blob/main/src/js/rollup.config.mjs#L41-L43
as an argument to terser which enables extra logging in debug builds. I don't see an analogue of that here. It looks to me like `DEBUG` is left undefined in the generated output so we'll see a CI failure here.

**Comment by WebReflection** [comment]
beside the *codespell* failing (no idea why) there's [no plugin](https://github.com/esbuild/community-plugins) to define a `DEBUG` reference in *esbuild* so I ended up creating my own thing that does the following:

  * if the content of the output contains a `DEBUG` reference
    * if the *format* is *IIFE* it makes it a default parameter value on the first `()` parens it finds
    * if the *format* is *ESM* it adds a `const DEBUG = ...;` reference before anything else in the module

I believe this satisfies previous requirements and, by visually checking files, it looks like everything should be fine now.

**Comment by WebReflection** [comment]
@hoodmane I hope tests will pass with the umd fake module but there's one thing we might need to address here: *source map*. Adding the `DEBUG` flag works well but it messes up information related to the minified file.

With the simplified umd dance I just append a little try/catch at the end so the rest of the source should be well mapped because that file doesn't use a `DEBUG` flag but the `api.ts` does use the flag and it'd feel weird to add it at the end, outside its *IIFE* scope. What do you think?

**Comment by WebReflection** [comment]
@hoodmane I've just pushed my last version that fixes pre-commits and should be green everywhere else (except one thing I don't know how to fix). I'll update the description and let you figure out how to fix that single CI thing. I hope the rest is fine/good.

**Comment by hoodmane** [comment]
Can you remove the unused `rollup.config.js` file?

**Comment by hoodmane** [comment]
Sorry for some reason I didn't see that it was removed.

File: docs/project/changelog.md [hunk 1]
```diff
@@ -15,6 +15,9 @@ myst:
 
 ## Unreleased
 
+- {{ Enhancement }} _esbuild_ is now used instead of _rollup_.
```
[the remark this answers is not in the document]
Comment (WebReflection) [reply docs/project/changelog.md]: Removed

**Comment by ryanking13** [comment]
Thanks, @WebReflection! We were relying on rollup to do the type check when build, and AFAIK esbuild does not do type checks. So perhaps we need to add type checks in pre-commit or somewhere else instead?

**Comment by WebReflection** [comment]
@ryanking13 you do have pre-commit checks for many things I'd be surprised TS checks are not applied there (via prettier or linter) but happy to add a `tsc` command in the build

**Comment by ryanking13** [comment]
> @ryanking13 you do have pre-commit checks for many things I'd be surprised TS checks are not applied there (via prettier or linter) but happy to add a tsc command in the build

I think we don't have one (maybe our prettier config is wrong), so I would appreciate it if you can include a type check in this PR.

**Comment by WebReflection** [comment]
@ryanking13 will do ... one extra question we discussed with @hoodmane yesterday ... does the config need to be compatible with Node.JS 14? I think for its task it shouldn't but I haven't tried it with legacy LTS versions, thanks.

**Comment by ryanking13** [comment]
> does the config need to be compatible with Node.JS 14?

I think it is okay to not support Node.js 14 as its end of life is coming soon. but there is an issue that emscripten uses node.js 14 internally, so I'm not very sure. So maybe let's make it compatible with Node.js 14 if it is not very hard?

**Comment by WebReflection** [comment]
@ryanking13 so I've tested with node 14 and everything seems to work as expected + I've added an explicit `tsc` script for `npm` to run and added that in the *Makefile* for building / checking all the things. Then I've changed a single `string` type to `number` and got an error indeed ... I hope this last change makes everyone happy.

**Review by hoodmane — APPROVED** [review]
Thanks again @WebReflection!

**Comment by hoodmane** [comment]
Oops I didn't notice that test-core-firefox is still failing...

**Comment by WebReflection** [comment]
@hoodmane yeah, that was the one you said you'll have a look 'cause I am not sure what is that about ... I should've reminded that in my last comment.

**Comment by hoodmane** [comment]
Well the issue is that in Chrome we get:
```js
{
    "columnNumber": 11,
    "lineNumber": 677,
    "fileName": "http://localhost:8000/pyodide.asm.js",
    "functionName": "new_error",
    "source": "    at new_error (http://localhost:8000/pyodide.asm.js:677:11)"
}
```
whereas in Firefox the same frame is:
```js
{
  "columnNumber": 13,
  "lineNumber": 5150,
  "fileName": "http://localhost:8000/pyodide.asm.js",
  "functionName": "S",
  "source": "S@http://localhost:8000/pyodide.asm.js:5150:13"
}
```
note that `functionName` is "S" and not `new_error`. We were relying on the function name not being mangled in the stack frame.

**Comment by hoodmane** [comment]
Okay so Chrome and Firefox were always generating different stack frames. Firefox points to PythonError:
```
{
  "columnNumber": 15,
  "lineNumber": 5278,
  "fileName": "http://localhost:8000/pyodide.asm.js",
  "functionName": "PythonError",
  "source": "PythonError@http://localhost:8000/pyodide.asm.js:5278:15"
}
```
For some reason `PythonError` gets mangled into `S` by esbuild whereas `new_error` is retained.

Incidentally, neither `new_error` nor `PythonError` really seem to me like they should be in the stack frame. I would think it would be:
```
hiwire_throw_error
pythonexc2js
... outer calls
```
but neither `hiwire_throw_error` nor `pythonexc2js` appear, and these frames that are involved in constructing the error appear instead.

**Pull request merged**